src/vmm/src/vstate/vm.rs

```diff
@@ -276,14 +499,215 @@ impl Vm {
         file.sync_all()
             .map_err(|err| MemoryBackingFile("sync_all", err))
     }
+
+    /// Register a device IRQ
+    pub fn register_irq(&self, fd: &EventFd, gsi: u32) -> Result<(), errno::Error> {
+        self.common.fd.register_irqfd(fd, gsi)?;
+
+        let mut entry = kvm_irq_routing_entry {
+            gsi,
+            type_: KVM_IRQ_ROUTING_IRQCHIP,
+            ..Default::default()
+        };
+        #[cfg(target_arch = "x86_64")]
+        {
+            entry.u.irqchip.irqchip = KVM_IRQCHIP_IOAPIC;
+        }
+        #[cfg(target_arch = "aarch64")]
+        {
+            entry.u.irqchip.irqchip = 0;
+        }
+        entry.u.irqchip.pin = gsi;
+
+        self.common
+            .interrupts
+            .lock()
+            .expect("Poisoned lock")
+            .insert(
+                gsi,
+                RoutingEntry {
+                    entry,
+                    masked: false,
+                },
+            );
+        Ok(())
+    }
+
+    /// Register an MSI device interrupt
+    pub fn register_msi(
+        &self,
+        route: &MsiVector,
+        masked: bool,
+        config: MsiIrqSourceConfig,
+    ) -> Result<(), errno::Error> {
+        let mut entry = kvm_irq_routing_entry {
+            gsi: route.gsi,
+            type_: KVM_IRQ_ROUTING_MSI,
+            ..Default::default()
+        };
+        entry.u.msi.address_lo = config.low_addr;
+        entry.u.msi.address_hi = config.high_addr;
+        entry.u.msi.data = config.data;
+
+        if self.common.fd.check_extension(kvm_ioctls::Cap::MsiDevid) {
+            // On AArch64, there is limitation on the range of the 'devid',
+            // it cannot be greater than 65536 (the max of u16).
+            //
+            // BDF cannot be used directly, because 'segment' is in high
+            // 16 bits. The layout of the u32 BDF is:
+            // |---- 16 bits ----|-- 8 bits --|-- 5 bits --|-- 3 bits --|
+            // |      segment    |     bus    |   device   |  function  |
+            //
+            // Now that we support 1 bus only in a segment, we can build a
+            // 'devid' by replacing the 'bus' bits with the low 8 bits of
+            // 'segment' data.
+            // This way we can resolve the range checking problem and give
+            // different `devid` to all the devices. Limitation is that at
+            // most 256 segments can be supported.
+            //
+            let modified_devid = ((config.devid & 0x00ff_0000) >> 8) | config.devid & 0xff;
+
+            entry.flags = KVM_MSI_VALID_DEVID;
+            entry.u.msi.__bindgen_anon_1.devid = modified_devid;
+        }
+
+        self.common
+            .interrupts
+            .lock()
+            .expect("Poisoned lock")
+            .insert(route.gsi, RoutingEntry { entry, masked });
+
+        Ok(())
+    }
+
+    /// Create a group of MSI-X interrupts
+    pub fn create_msix_group(
+        vm: Arc<Vm>,
+        base: u32,
```

Comment:
As I can see, the base is always 0. So maybe just remove this param?

Reply:
done